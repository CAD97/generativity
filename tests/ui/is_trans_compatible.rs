// see cad97/generativity#13 and rust-lang/rust#78586
// if this fails in crater *please* ping me; I *want* this to not hit the lint!
#![deny(repr_transparent_external_private_fields)] // this should not trigger

use generativity::{make_guard, Id};

#[repr(transparent)]
pub struct BOption<'id, T>(Option<T>, Id<'id>); // this should work

fn main() {
    make_guard!(a);
    let _ = BOption(Some(0), a.into());
}
