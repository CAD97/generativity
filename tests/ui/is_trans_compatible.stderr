error: zero-sized fields in `repr(transparent)` cannot contain external non-exhaustive types
 --> tests/ui/is_trans_compatible.rs:8:39
  |
8 | pub struct BOption<'id, T>(Option<T>, Id<'id>); // this should work
  |                                       ^^^^^^^
  |
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #78586 <https://github.com/rust-lang/rust/issues/78586>
  = note: this struct contains `Id<'id>`, which contains private fields, and makes it not a breaking change to become non-zero-sized in the future.
note: the lint level is defined here
 --> tests/ui/is_trans_compatible.rs:3:9
  |
3 | #![deny(repr_transparent_external_private_fields)] // this should not trigger
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
